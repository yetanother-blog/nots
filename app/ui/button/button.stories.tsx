import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './button';

const meta: Meta<typeof Button> = {
  title: 'Forms/Button',
  component: Button,
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    children: 'Primary Button',
  },
};
