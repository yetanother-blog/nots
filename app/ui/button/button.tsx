import React from "react";
import { clsx } from "clsx/lite";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {}

export const Button: React.FC<ButtonProps> = ({
  children,
  className,
  ...props
}) => {
  const classes = clsx("rounded bg-blue-500 text-white px-4 py-3", className);

  return (
    <button className={classes} {...props}>
      {children}
    </button>
  );
};
