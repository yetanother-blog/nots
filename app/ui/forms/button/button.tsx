import { forwardRef } from 'react';
import { clsx } from 'clsx';
import { Typography } from '~/ui/data-display/typography/typography';
import type { IconName } from '~/ui';
import { LoadingIndicator, Icon } from '~/ui';
import type { PolymorphicComponentPropWithRef } from '~/ui/utils/polymorphic-types';

type Variant = 'filled';
export type ButtonColor = 'grey';
type Size = 'regular';

export type ButtonProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<
    C,
    {
      isLoading?: boolean;
      variant?: Variant;
      color?: ButtonColor;
      size?: Size;
      iconPosition?: 'start' | 'end';
      icon?: IconName;
      className?: string;
    }
  >;

type ButtonComponent = <C extends React.ElementType = 'button'>(
  props: ButtonProps<C>
) => React.ReactElement | null;

// eslint-disable-next-line react/display-name
export const Button = forwardRef(
  <C extends React.ElementType = 'button'>(
    {
      as,
      children,
      isLoading,
      disabled,
      color = 'grey',
      size = 'regular',
      variant = 'filled',
      icon,
      iconPosition = 'end',
      className,
      ...props
    }: ButtonProps<C>,
    ref?: unknown
  ) => {
    /*
     * Filled Variants
     */
    const filledClasses = clsx([
      'text-white dark:text-nots-grey-800',
      isLoading &&
        'drop-shadow-none hover:drop-shadow-none pointer-events-none',
    ]);

    const primaryFilledClasses = clsx([
      filledClasses,
      'disabled:bg-nots-grey-200 dark:disabled:bg-white dark:disabled:text-nots-grey-200 focus:outline-none',
      'bg-nots-grey-800 dark:bg-white hover:bg-nots-grey-600 dark:hover:bg-nots-grey-100 focus:outline-none focus:ring-2 focus:ring-nots-grey-300',
      isLoading && 'hover:bg-ws-primary-500',
    ]);

    const colorClasses = clsx([
      variant === 'filled' && color === 'grey' && primaryFilledClasses,
    ]);

    const baseShapeClasses = clsx([size === 'regular' && `px-3 py-[6px]`]);

    const iconOnlyClasses = clsx(['w-9 h-8', 'py-0', 'px-0']);
    const isOnlyIcon = !children && icon;

    const classes = clsx([
      'flex items-center justify-center rounded-sm',
      'transition-colors duration-200 ease-in-out',
      'gap-2',
      isOnlyIcon ? iconOnlyClasses : baseShapeClasses,
      iconPosition === 'start' && 'flex-row-reverse',
      iconPosition === 'end' && 'flex-row',
      isLoading && 'pointer-events-none',
      colorClasses,
      className,
    ]);

    const hideClasses = clsx([isLoading && 'opacity-0 pointer-events-none']);

    const Component = as || 'button';

    return (
      <Component className={classes} disabled={disabled} {...props} ref={ref}>
        {children && (
          <Typography
            className={hideClasses}
            variant="heading3"
            fontWeight="semibold"
            as="span"
          >
            {children}
          </Typography>
        )}
        {isLoading && <LoadingIndicator className="absolute" />}
        {icon && <Icon icon={icon} className={hideClasses} />}
      </Component>
    );
  }
) as ButtonComponent;
