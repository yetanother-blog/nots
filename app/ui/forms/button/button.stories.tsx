import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './button';

/**
 * The Button component is used to trigger an action or event, such as
 * submitting a form, opening a dialog, canceling an action, or performing a
 * delete operation.
 *
 * # Render as a Remix Router Link
 *
 * It's also possible to use the Button in combination with Remix Router (aka
 * React Router). Just pass the component of the routing library to the UI
 * component.
 *
 * ```tsx
 * import { Link } from '@remix-run/react';
 *
 * ...
 *
 * <Button as={Link} to="/some-route">
 *  Remix Router Link
 * </Button>
 * ```
 */
const meta: Meta<typeof Button> = {
  title: 'Forms/Button',
  component: Button,
  argTypes: {
    color: {
      control: {
        type: 'select',
        options: ['grey'],
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const GreyFilled: Story = {
  args: {
    color: 'grey',
  },
  render: () => (
    <div className="flex flex-grow-0 flex-wrap gap-4">
      <Button onClick={() => console.log('clicked')}>Primary Button</Button>
      <Button isLoading={true} onClick={() => console.log('clicked')}>
        Primary Button
      </Button>
      <Button disabled={true} onClick={() => console.log('clicked')}>
        Primary Button
      </Button>
      <Button
        onClick={() => console.log('clicked')}
        icon="caretDown"
        iconPosition="start"
      >
        Primary Button
      </Button>
      <Button onClick={() => console.log('clicked')} icon="caretDown">
        Primary Button
      </Button>
      <Button
        isLoading={true}
        onClick={() => console.log('clicked')}
        icon="caretDown"
      >
        Primary Button
      </Button>
      <Button
        disabled={true}
        onClick={() => console.log('clicked')}
        icon="paragraph"
      >
        Primary Button
      </Button>
      <Button onClick={() => console.log('clicked')} icon="paragraph" />
      <Button
        isLoading={true}
        onClick={() => console.log('clicked')}
        icon="paragraph"
      />
      <Button
        disabled={true}
        onClick={() => console.log('clicked')}
        icon="paragraph"
      />
    </div>
  ),
};
