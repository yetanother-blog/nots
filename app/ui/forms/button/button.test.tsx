import { render } from '@testing-library/react';
import type { ButtonColor } from './button';
import { Button } from './button';

describe('Button', () => {
  it.each<[ButtonColor, string]>([['grey', 'bg-nots-grey-800']])(
    'should render %s filled button',
    (variant, expectedColor) => {
      const { getByRole, container } = render(
        <Button variant="filled" color={variant} />
      );
      const button = getByRole('button');
      expect(button).toHaveClass(expectedColor);
      expect(container).toMatchSnapshot();
    }
  );

  it('should render disabled button', () => {
    const { getByRole, container } = render(
      <Button variant="filled" disabled />
    );
    const button = getByRole('button');
    expect(button).toHaveClass('disabled:bg-nots-grey-200');
    expect(container).toMatchSnapshot();
  });

  it('should render loading button', () => {
    const { getByRole, container } = render(
      <Button variant="filled" isLoading />
    );
    const loadingIndicator = getByRole('status');

    expect(loadingIndicator).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render button with custom class', () => {
    const { getByRole } = render(
      <Button variant="filled" className="custom-class" />
    );
    const button = getByRole('button');
    expect(button).toHaveClass('custom-class');
  });

  it('should render button as a link', () => {
    const { getByRole } = render(
      <Button variant="filled" as="a" href="https://example.com" />
    );
    const button = getByRole('link');
    expect(button).toHaveAttribute('href', 'https://example.com');
  });

  it('should render button with icon', () => {
    const { getByRole } = render(<Button variant="filled" icon="caretDown" />);
    const button = getByRole('button');
    expect(button.innerHTML).toContain('svg');
  });
});
