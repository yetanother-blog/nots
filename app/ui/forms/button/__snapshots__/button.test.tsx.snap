// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render disabled button 1`] = `
<div>
  <button
    class="flex items-center justify-center rounded-sm transition-colors duration-200 ease-in-out gap-2 px-3 py-[6px] flex-row text-white dark:text-nots-grey-800 disabled:bg-nots-grey-200 dark:disabled:bg-white dark:disabled:text-nots-grey-200 focus:outline-none bg-nots-grey-800 dark:bg-white hover:bg-nots-grey-600 dark:hover:bg-nots-grey-100 focus:outline-none focus:ring-2 focus:ring-nots-grey-300"
    disabled=""
  />
</div>
`;

exports[`Button > should render grey filled button 1`] = `
<div>
  <button
    class="flex items-center justify-center rounded-sm transition-colors duration-200 ease-in-out gap-2 px-3 py-[6px] flex-row text-white dark:text-nots-grey-800 disabled:bg-nots-grey-200 dark:disabled:bg-white dark:disabled:text-nots-grey-200 focus:outline-none bg-nots-grey-800 dark:bg-white hover:bg-nots-grey-600 dark:hover:bg-nots-grey-100 focus:outline-none focus:ring-2 focus:ring-nots-grey-300"
  />
</div>
`;

exports[`Button > should render loading button 1`] = `
<div>
  <button
    class="flex items-center justify-center rounded-sm transition-colors duration-200 ease-in-out gap-2 px-3 py-[6px] flex-row pointer-events-none text-white dark:text-nots-grey-800 drop-shadow-none hover:drop-shadow-none pointer-events-none disabled:bg-nots-grey-200 dark:disabled:bg-white dark:disabled:text-nots-grey-200 focus:outline-none bg-nots-grey-800 dark:bg-white hover:bg-nots-grey-600 dark:hover:bg-nots-grey-100 focus:outline-none focus:ring-2 focus:ring-nots-grey-300 hover:bg-ws-primary-500"
  >
    <div
      class="pointer-events-none flex items-center justify-center absolute"
    >
      <svg
        aria-label="loading"
        class="text-current inline-block h-6 w-6 animate-spin"
        fill="transparent"
        focusable="false"
        height="24"
        role="status"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <span
          class="sr-only"
        >
          Loading...
        </span>
        <path
          d="M12 2V6M12 18V22M6 12H2M22 12H18M19.0784 19.0784L16.25 16.25M19.0784 4.99994L16.25 7.82837M4.92157 19.0784L7.75 16.25M4.92157 4.99994L7.75 7.82837"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
  </button>
</div>
`;
